File: python/graph_general/number_islands.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0

        num_islands = 0
        rows, cols = len(grid), len(grid[0])

        def dfs(r, c):
            if r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] == '0':
                return

            grid[r][c] = '0'

            dfs(r + 1, c)  # bawah
            dfs(r - 1, c)  # atas
            dfs(r, c + 1)  # kanan
            dfs(r, c - 1)  # kiri

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '1':
                    num_islands += 1

                    dfs(r,c)

        return num_islands

File: golang/graph_general/number_islands.go
package main

func numIslands(grid [][]byte) int {
	if len(grid) == 0 {
		return 0
	}

	numIslands := 0

	rows := len(grid)
	cols := len(grid[0])

	var dfs func(r, c int)

	dfs = func(r, c int) {
		if r < 0 || r >= rows || c < 0 || c >= cols || grid[r][c] == '0' {
			return
		}

		grid[r][c] = '0'

		dfs(r+1, c) // bawah
		dfs(r-1, c) // atas
		dfs(r, c+1) // kanan
		dfs(r, c-1) // kiri
	}

	for r := 0; r < rows; r++ {

		for c := 0; c < cols; c++ {
			if grid[r][c] == '1' {
				numIslands++
				dfs(r, c)
			}
		}
	}

	return numIslands
}
